From 5c38d3d38d174fc2e68d97f30e44b17cd3c365f8 Mon Sep 17 00:00:00 2001
From: Roman Birg <roman@cyngn.com>
Date: Thu, 4 Feb 2016 16:28:00 -0800
Subject: [PATCH 50/77] framework: move device key handler logic, fix gesture
 camera launch

In order to make device camera gestures behave the same way as launching
the camera by double tapping the power button, route DeviceKeyHandler
logic through the GestureLauncherService.

A callback is used in order to avoid publishing the
GestureLauncherService into the System service registry, because the
KeyHandler code cannot directly access services local to the system
process.

OPO-427

Change-Id: I2a8860b278f65ccf31bfee151513d8a23ab746c5
Signed-off-by: Roman Birg <roman@cyngn.com>

KeyHandler: don't break old interfaces

Causing bootloops on devices, don't break old interfaces in the
framework.

OPO-427

Change-Id: I49473b057bc4ed4710e9092c76e41c6a8134d2de
Signed-off-by: Roman Birg <roman@cyngn.com>

SystemUI: handle camera launch gesture from keyhadler

Includes partial reverts:

    Revert "KeyHandler: don't break old interfaces"
    This reverts commit 31ecbdc5fcc1263e9028107858565d951798fd93.

    Revert "framework: move device key handler logic, fix gesture camera launch"
    This reverts commit db1b4e6e4880be1828342be9bc468d36565a2ef0.

Ref: OPO-427

Change-Id: Ic15543a8f63ad4521ab0ab536d00224b0bd70f8c
Signed-off-by: Roman Birg <roman@cyngn.com>
---
 core/java/android/app/StatusBarManager.java         |  1 +
 .../android/systemui/statusbar/phone/StatusBar.java | 13 +++++++++++++
 2 files changed, 14 insertions(+)

diff --git a/core/java/android/app/StatusBarManager.java b/core/java/android/app/StatusBarManager.java
index b83b44d295b..8af64b63681 100644
--- a/core/java/android/app/StatusBarManager.java
+++ b/core/java/android/app/StatusBarManager.java
@@ -106,6 +106,7 @@ public class StatusBarManager {
     public static final int CAMERA_LAUNCH_SOURCE_WIGGLE = 0;
     public static final int CAMERA_LAUNCH_SOURCE_POWER_DOUBLE_TAP = 1;
     public static final int CAMERA_LAUNCH_SOURCE_LIFT_TRIGGER = 2;
+    public static final int CAMERA_LAUNCH_SOURCE_SCREEN_GESTURE = 3;
 
     private Context mContext;
     private IStatusBarService mService;
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/phone/StatusBar.java b/packages/SystemUI/src/com/android/systemui/statusbar/phone/StatusBar.java
index f19da1252e1..f176a3b5125 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/phone/StatusBar.java
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/phone/StatusBar.java
@@ -1101,6 +1101,7 @@ public class StatusBar extends SystemUI implements DemoMode,
         filter.addAction(Intent.ACTION_CLOSE_SYSTEM_DIALOGS);
         filter.addAction(Intent.ACTION_SCREEN_OFF);
         filter.addAction(DevicePolicyManager.ACTION_SHOW_DEVICE_MONITORING_DIALOG);
+        filter.addAction(lineageos.content.Intent.ACTION_SCREEN_CAMERA_GESTURE);
         context.registerReceiverAsUser(mBroadcastReceiver, UserHandle.ALL, filter, null, null);
 
         IntentFilter demoFilter = new IntentFilter();
@@ -3099,6 +3100,18 @@ public class StatusBar extends SystemUI implements DemoMode,
             else if (DevicePolicyManager.ACTION_SHOW_DEVICE_MONITORING_DIALOG.equals(action)) {
                 mQSPanel.showDeviceMonitoringDialog();
             }
+            else if (lineageos.content.Intent.ACTION_SCREEN_CAMERA_GESTURE.equals(action)) {
+                boolean userSetupComplete = Settings.Secure.getInt(mContext.getContentResolver(),
+                        Settings.Secure.USER_SETUP_COMPLETE, 0) != 0;
+                if (!userSetupComplete) {
+                    if (DEBUG) Log.d(TAG, String.format(
+                            "userSetupComplete = %s, ignoring camera launch gesture.",
+                            userSetupComplete));
+                    return;
+                }
+
+                onCameraLaunchGestureDetected(StatusBarManager.CAMERA_LAUNCH_SOURCE_SCREEN_GESTURE);
+            }
         }
     };
 
-- 
2.17.1

